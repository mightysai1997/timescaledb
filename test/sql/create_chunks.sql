\ir include/create_single_db.sql

CREATE TABLE chunk_test(
        time       BIGINT,
        metric     INTEGER,
        device_id  TEXT
    );

-- Test chunk closing/creation
SELECT * FROM create_hypertable('chunk_test', 'time', 'device_id', 2, chunk_time_interval => 10);
SELECT * FROM _timescaledb_catalog.hypertable;

INSERT INTO chunk_test VALUES (1, 1, 'dev1'),
                              (2, 2, 'dev2'),
                              (45, 2, 'dev2'),
                              (46, 2, 'dev2');

SELECT * FROM set_chunk_time_interval('chunk_test', 40::bigint);


INSERT INTO chunk_test VALUES(23, 3, 'dev3');

SELECT * FROM chunk_test order by time, metric, device_id;
SELECT * FROM _timescaledb_catalog.chunk;
SELECT * FROM _timescaledb_catalog.hypertable;

SELECT * FROM ONLY chunk_test;
SELECT * FROM _timescaledb_catalog.chunk c
    LEFT JOIN _timescaledb_catalog.partition p ON (p.id = c.partition_id)
    LEFT JOIN _timescaledb_catalog.partition_epoch pe ON (pe.id = p.epoch_id)
    LEFT JOIN _timescaledb_catalog.hypertable h ON (pe.hypertable_id = h.id)
    WHERE h.schema_name = 'public' AND h.table_name = 'chunk_test'
    ORDER BY c.id;

