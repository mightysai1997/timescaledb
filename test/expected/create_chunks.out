\ir include/create_single_db.sql
SET client_min_messages = WARNING;
DROP DATABASE IF EXISTS single;
SET client_min_messages = NOTICE;
CREATE DATABASE single;
\c single
CREATE EXTENSION IF NOT EXISTS timescaledb;
SET timescaledb.disable_optimizations = :DISABLE_OPTIMIZATIONS;
CREATE TABLE chunk_test(
        time       BIGINT,
        metric     INTEGER,
        device_id  TEXT
    );
-- Test chunk closing/creation
SELECT * FROM create_hypertable('chunk_test', 'time', 'device_id', 2, chunk_time_interval => 10);
 create_hypertable 
-------------------
 
(1 row)

SELECT * FROM _timescaledb_catalog.hypertable;
 id | schema_name | table_name | associated_schema_name | associated_table_prefix | num_dimensions 
----+-------------+------------+------------------------+-------------------------+----------------
  1 | public      | chunk_test | _timescaledb_internal  | _hyper_1                |              2
(1 row)

INSERT INTO chunk_test VALUES (1, 1, 'dev1'),
                              (2, 2, 'dev2'),
                              (45, 2, 'dev2'),
                              (46, 2, 'dev2');
SELECT * FROM set_chunk_time_interval('chunk_test', 40::bigint);
 set_chunk_time_interval 
-------------------------
 
(1 row)

INSERT INTO chunk_test VALUES(23, 3, 'dev3');
SELECT * FROM chunk_test order by time, metric, device_id;
 time | metric | device_id 
------+--------+-----------
    1 |      1 | dev1
    2 |      2 | dev2
   23 |      3 | dev3
   45 |      2 | dev2
   46 |      2 | dev2
(5 rows)

SELECT * FROM _timescaledb_catalog.chunk;
 id | hypertable_id |      schema_name      |    table_name    
----+---------------+-----------------------+------------------
  1 |             1 | _timescaledb_internal | _hyper_1_1_chunk
  2 |             1 | _timescaledb_internal | _hyper_1_2_chunk
  3 |             1 | _timescaledb_internal | _hyper_1_3_chunk
  4 |             1 | _timescaledb_internal | _hyper_1_4_chunk
(4 rows)

SELECT * FROM _timescaledb_catalog.hypertable;
 id | schema_name | table_name | associated_schema_name | associated_table_prefix | num_dimensions 
----+-------------+------------+------------------------+-------------------------+----------------
  1 | public      | chunk_test | _timescaledb_internal  | _hyper_1                |              2
(1 row)

SELECT * FROM ONLY chunk_test;
 time | metric | device_id 
------+--------+-----------
(0 rows)

SELECT * FROM _timescaledb_catalog.chunk c
    LEFT JOIN _timescaledb_catalog.chunk_constraint cc ON (c.id = cc.chunk_id)
    LEFT JOIN _timescaledb_catalog.dimension_slice ds ON (ds.id = cc.dimension_slice_id)
    LEFT JOIN _timescaledb_catalog.dimension d ON (d.id = ds.dimension_id)
    LEFT JOIN _timescaledb_catalog.hypertable h ON (d.hypertable_id = h.id)
    WHERE h.schema_name = 'public' AND h.table_name = 'chunk_test'
    ORDER BY c.id, d.id;
 id | hypertable_id |      schema_name      |    table_name    | chunk_id | dimension_slice_id | constraint_name | hypertable_constraint_name | id | dimension_id | range_start | range_end  | id | hypertable_id | column_name | column_type | aligned | num_slices | partitioning_func_schema |   partitioning_func   | interval_length | id | schema_name | table_name | associated_schema_name | associated_table_prefix | num_dimensions 
----+---------------+-----------------------+------------------+----------+--------------------+-----------------+----------------------------+----+--------------+-------------+------------+----+---------------+-------------+-------------+---------+------------+--------------------------+-----------------------+-----------------+----+-------------+------------+------------------------+-------------------------+----------------
  1 |             1 | _timescaledb_internal | _hyper_1_1_chunk |        1 |                  1 | constraint_1    |                            |  1 |            1 |           0 |         10 |  1 |             1 | time        | bigint      | t       |            |                          |                       |              40 |  1 | public      | chunk_test | _timescaledb_internal  | _hyper_1                |              2
  1 |             1 | _timescaledb_internal | _hyper_1_1_chunk |        1 |                  2 | constraint_2    |                            |  2 |            2 |  1073741823 | 2147483647 |  2 |             1 | device_id   | text        | f       |          2 | _timescaledb_internal    | get_partition_for_key |                 |  1 | public      | chunk_test | _timescaledb_internal  | _hyper_1                |              2
  2 |             1 | _timescaledb_internal | _hyper_1_2_chunk |        2 |                  1 | constraint_1    |                            |  1 |            1 |           0 |         10 |  1 |             1 | time        | bigint      | t       |            |                          |                       |              40 |  1 | public      | chunk_test | _timescaledb_internal  | _hyper_1                |              2
  2 |             1 | _timescaledb_internal | _hyper_1_2_chunk |        2 |                  4 | constraint_4    |                            |  4 |            2 |           0 | 1073741823 |  2 |             1 | device_id   | text        | f       |          2 | _timescaledb_internal    | get_partition_for_key |                 |  1 | public      | chunk_test | _timescaledb_internal  | _hyper_1                |              2
  3 |             1 | _timescaledb_internal | _hyper_1_3_chunk |        3 |                  5 | constraint_5    |                            |  5 |            1 |          40 |         50 |  1 |             1 | time        | bigint      | t       |            |                          |                       |              40 |  1 | public      | chunk_test | _timescaledb_internal  | _hyper_1                |              2
  3 |             1 | _timescaledb_internal | _hyper_1_3_chunk |        3 |                  4 | constraint_4    |                            |  4 |            2 |           0 | 1073741823 |  2 |             1 | device_id   | text        | f       |          2 | _timescaledb_internal    | get_partition_for_key |                 |  1 | public      | chunk_test | _timescaledb_internal  | _hyper_1                |              2
  4 |             1 | _timescaledb_internal | _hyper_1_4_chunk |        4 |                  7 | constraint_7    |                            |  7 |            1 |          10 |         40 |  1 |             1 | time        | bigint      | t       |            |                          |                       |              40 |  1 | public      | chunk_test | _timescaledb_internal  | _hyper_1                |              2
  4 |             1 | _timescaledb_internal | _hyper_1_4_chunk |        4 |                  4 | constraint_4    |                            |  4 |            2 |           0 | 1073741823 |  2 |             1 | device_id   | text        | f       |          2 | _timescaledb_internal    | get_partition_for_key |                 |  1 | public      | chunk_test | _timescaledb_internal  | _hyper_1                |              2
(8 rows)

-- Test chunk aligning between partitions
CREATE TABLE chunk_align_test(
        time       BIGINT,
        metric     INTEGER,
        device_id  TEXT
    );
SELECT * FROM create_hypertable('chunk_align_test', 'time', 'device_id', 2, chunk_time_interval => 10);
 create_hypertable 
-------------------
 
(1 row)

INSERT INTO chunk_align_test VALUES (1, 1, 'dev1'); -- this should create a 10 wide chunk
SELECT * FROM _timescaledb_catalog.chunk c
    LEFT JOIN _timescaledb_catalog.chunk_constraint cc ON (c.id = cc.chunk_id)
    LEFT JOIN _timescaledb_catalog.dimension_slice ds ON (ds.id = cc.dimension_slice_id)
    LEFT JOIN _timescaledb_catalog.dimension d ON (d.id = ds.dimension_id)
    LEFT JOIN _timescaledb_catalog.hypertable h ON (d.hypertable_id = h.id)
    WHERE h.schema_name = 'public' AND h.table_name = 'chunk_align_test'
          AND d.column_name = 'time'
    ORDER BY c.id, d.id;
 id | hypertable_id |      schema_name      |    table_name    | chunk_id | dimension_slice_id | constraint_name | hypertable_constraint_name | id | dimension_id | range_start | range_end | id | hypertable_id | column_name | column_type | aligned | num_slices | partitioning_func_schema | partitioning_func | interval_length | id | schema_name |    table_name    | associated_schema_name | associated_table_prefix | num_dimensions 
----+---------------+-----------------------+------------------+----------+--------------------+-----------------+----------------------------+----+--------------+-------------+-----------+----+---------------+-------------+-------------+---------+------------+--------------------------+-------------------+-----------------+----+-------------+------------------+------------------------+-------------------------+----------------
  5 |             2 | _timescaledb_internal | _hyper_2_5_chunk |        5 |                  9 | constraint_9    |                            |  9 |            3 |           0 |        10 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              10 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
(1 row)

    
SELECT * FROM set_chunk_time_interval('chunk_align_test', 40::bigint);
 set_chunk_time_interval 
-------------------------
 
(1 row)

INSERT INTO chunk_align_test VALUES (5, 1, 'dev2'); -- this should still create a 10 wide chunk
INSERT INTO chunk_align_test VALUES (45, 1, 'dev2'); -- this should create a 40 wide chunk
SELECT * FROM _timescaledb_catalog.chunk c
    LEFT JOIN _timescaledb_catalog.chunk_constraint cc ON (c.id = cc.chunk_id)
    LEFT JOIN _timescaledb_catalog.dimension_slice ds ON (ds.id = cc.dimension_slice_id)
    LEFT JOIN _timescaledb_catalog.dimension d ON (d.id = ds.dimension_id)
    LEFT JOIN _timescaledb_catalog.hypertable h ON (d.hypertable_id = h.id)
    WHERE h.schema_name = 'public' AND h.table_name = 'chunk_align_test'
          AND d.column_name = 'time'
    ORDER BY c.id, d.id;
 id | hypertable_id |      schema_name      |    table_name    | chunk_id | dimension_slice_id | constraint_name | hypertable_constraint_name | id | dimension_id | range_start | range_end | id | hypertable_id | column_name | column_type | aligned | num_slices | partitioning_func_schema | partitioning_func | interval_length | id | schema_name |    table_name    | associated_schema_name | associated_table_prefix | num_dimensions 
----+---------------+-----------------------+------------------+----------+--------------------+-----------------+----------------------------+----+--------------+-------------+-----------+----+---------------+-------------+-------------+---------+------------+--------------------------+-------------------+-----------------+----+-------------+------------------+------------------------+-------------------------+----------------
  5 |             2 | _timescaledb_internal | _hyper_2_5_chunk |        5 |                  9 | constraint_9    |                            |  9 |            3 |           0 |        10 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              40 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
  6 |             2 | _timescaledb_internal | _hyper_2_6_chunk |        6 |                  9 | constraint_9    |                            |  9 |            3 |           0 |        10 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              40 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
  7 |             2 | _timescaledb_internal | _hyper_2_7_chunk |        7 |                 13 | constraint_13   |                            | 13 |            3 |          40 |        80 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              40 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
(3 rows)

    --check the cut-to-size with aligned dimensions code
    INSERT INTO chunk_align_test VALUES (35, 1, 'dev1');
    INSERT INTO chunk_align_test VALUES (35, 1, 'dev2');
    INSERT INTO chunk_align_test VALUES (81, 1, 'dev1');
    INSERT INTO chunk_align_test VALUES (81, 1, 'dev2');
SELECT * FROM _timescaledb_catalog.chunk c
    LEFT JOIN _timescaledb_catalog.chunk_constraint cc ON (c.id = cc.chunk_id)
    LEFT JOIN _timescaledb_catalog.dimension_slice ds ON (ds.id = cc.dimension_slice_id)
    LEFT JOIN _timescaledb_catalog.dimension d ON (d.id = ds.dimension_id)
    LEFT JOIN _timescaledb_catalog.hypertable h ON (d.hypertable_id = h.id)
    WHERE h.schema_name = 'public' AND h.table_name = 'chunk_align_test'
          AND d.column_name = 'time'
    ORDER BY c.id, d.id;
 id | hypertable_id |      schema_name      |    table_name     | chunk_id | dimension_slice_id | constraint_name | hypertable_constraint_name | id | dimension_id | range_start | range_end | id | hypertable_id | column_name | column_type | aligned | num_slices | partitioning_func_schema | partitioning_func | interval_length | id | schema_name |    table_name    | associated_schema_name | associated_table_prefix | num_dimensions 
----+---------------+-----------------------+-------------------+----------+--------------------+-----------------+----------------------------+----+--------------+-------------+-----------+----+---------------+-------------+-------------+---------+------------+--------------------------+-------------------+-----------------+----+-------------+------------------+------------------------+-------------------------+----------------
  5 |             2 | _timescaledb_internal | _hyper_2_5_chunk  |        5 |                  9 | constraint_9    |                            |  9 |            3 |           0 |        10 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              40 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
  6 |             2 | _timescaledb_internal | _hyper_2_6_chunk  |        6 |                  9 | constraint_9    |                            |  9 |            3 |           0 |        10 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              40 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
  7 |             2 | _timescaledb_internal | _hyper_2_7_chunk  |        7 |                 13 | constraint_13   |                            | 13 |            3 |          40 |        80 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              40 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
  8 |             2 | _timescaledb_internal | _hyper_2_8_chunk  |        8 |                 15 | constraint_15   |                            | 15 |            3 |          10 |        40 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              40 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
  9 |             2 | _timescaledb_internal | _hyper_2_9_chunk  |        9 |                 15 | constraint_15   |                            | 15 |            3 |          10 |        40 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              40 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
 10 |             2 | _timescaledb_internal | _hyper_2_10_chunk |       10 |                 19 | constraint_19   |                            | 19 |            3 |          80 |       120 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              40 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
 11 |             2 | _timescaledb_internal | _hyper_2_11_chunk |       11 |                 19 | constraint_19   |                            | 19 |            3 |          80 |       120 |  3 |             2 | time        | bigint      | t       |            |                          |                   |              40 |  2 | public      | chunk_align_test | _timescaledb_internal  | _hyper_2                |              2
(7 rows)

